Ticket2653 no sensor disables eurotherm ramping

ThomasLohnert opened the pull request:
### Description of work

Ramping on the Eurotherm does not work while no sensor is attached to the device, and is preventing users from setting a setpoint. With these changes, ramping is automatically disabled when the sensor is disconnected. The previous ramp setting is cached, and restored once the sensor is reconnected.

There is a physical device in the IBEX office that can be used for testing.

Related PRs: 
https://github.com/ISISComputingGroup/ibex_gui/pull/588
https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/41

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2653

### Acceptance criteria

* Can toggle `Ramp` while sensor is connected to device
* Can not toggle `Ramp` while sensor is disconnected from device
* `Ramp` is off while sensor is disconnected
* Can set temperature setpoint when sensor is disconnected
* Reconnecting the sensor restores the previous `Ramp` setting

---

#### Code Review

- [ ] **Pertitent information and a copy of the manual has been stored in the [wiki](IOCs)**
- [x] Does the IOC conform to IBEX standards?
    - [PV naming](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
    - [Disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records)
    - [Record simulation](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation)
    - [Finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches)
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)?
- [x] Have the changes been recorded appropriately in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Functional Tests

- IOC responds correctly in:
    - [ ] Devsim mode
    - [ ] Recsim mode
    - [x] Real device, if available
- [x] Supplementary IOCs (`..._0n` where `n>1`) run correctly
- [x] Log files do not report undefined macros (serach for `macLib: macro` to find instances of `macLib: macro [macro name] is undefined...`

### Final steps

- [x] Update the IOC submodule in the main EPICS repo. See [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

AdrianPotter commented on EUROTHRM/EUROTHRM-IOC-01App/Db/devEurothrm.db:
```diff
@@ -224,6 +229,83 @@ record(ai, "$(P)OUT_MAX")
+{
+   field(DESC, "Check whether the temperature sensor is reconnected")
+   field(INPA, "$(P)TEMP")
+   field(CALC, "A = 1529")
```
Where does this magic number `1529` come from? Can you add a comment. Probably also better to hold it in a PV so we can change it more easily and it isn't duplicated here and in `SENSOR_DISCONNECTED`

ThomasLohnert replied:
1529 is the precise temperature value the eurotherm returns when no sensor is detected. I will add a comment for that

AdrianPotter commented on EUROTHRM/EUROTHRM-IOC-01App/Db/devEurothrm.db:
```diff
@@ -224,6 +229,83 @@ record(ai, "$(P)OUT_MAX")
+{
+   field(LNK1, "$(P)SENSOR_CONNECTED")
+   field(LNK2, "$(P)SENSOR_DISCONNECTED")
+}
```
Can you get rid of this record if you set the `INPA` field in `SENSOR_(DIS)CONNECTED` to be `CP`?

AdrianPotter commented on EUROTHRM/EUROTHRM-IOC-01App/Db/devEurothrm.db:
```diff
@@ -146,13 +146,16 @@ record(ao, "$(P)CURRENT_TEMP")
 	field(DTYP, "asynFloat64")
 	field(OUT, "@asyn($(READ),0,1)CUR")
 	field(EGU, "K")
+    field(FLNK, "$(P)SENSOR_CHECK")
```
As mentioned on `SENSOR_CHECK`, `CP` would potentially be neater to avoid the `FANOUT`

ThomasLohnert commented:
Reworked to use `RBV` instead of `TEMP` to detect a disconnected sensor in order to work independently of calibration files (in addition to addressing above comments)
